### ISSUE TITLE
The problem of separating foreground/background

### ISSUE THREAD

**Issue by kehuantiantang** [daviswer/fewshotlocal#3]
Thank you so much for sharing your code to us. But I still a little confuse about how to separate the foreground/backgound in method `fl`.
```
b = inp.size(0)
# Create folded bcentroids, mask off inputs
# (B, 2, 10, 10) --> (B, 1, 2, 10, 10)
masks = torch.cat([m, 1 - m], dim=1).view(b, 1, 2, m.size(2), m.size(3))  # B 1 2 10 10

# (B, 1, 2, 10, 10) --> view(B, 1, 2, 10*10) --> sum(B, 1, 2)
bsize = masks.view(*masks.size()[:-2], -1).sum(-1) + .01  # B 1 2

# This operate separate the fore/backgound and normalizate it.
bcentroids = (inp.unsqueeze(2) * masks).view(b, inp.size(1), 2, -1).sum(-1) / bsize  # B 64 2
```

Why we need following code to create a centroid and make multiply operation with origin input value ?
```
bcentroids = (bcentroids.sum(0).unsqueeze(0) - bcentroids).unsqueeze(-1).unsqueeze(-1) / (b - 1)  # B 64 2 1 1


masks = torch.sum((inp.unsqueeze(2) - bcentroids) ** 2, 1).neg().unsqueeze(1)  # B 1 2 10 10
masks = F.softmax(masks, dim=2)


# Perform fore/back separation and appropriate expansion
# view(B, 1, 2, 100) --> sum(B, 1, 2) + 0.01
bsize = masks.view(*masks.size()[:-2], -1).sum(-1) + .01

# [b, 64, 1， 10, 10] / (B, 1, 2, 10, 10) --> (B 64 2 100)
out = (inp.unsqueeze(2) * masks).view(b, inp.size(1), 2, -1)  # B 64 2 100
```
Hope you can give me so help, thank you so much !

**Comment by daviswer** [comment]
Hello! The purpose of the remaining code is to perform batch folding on the box centroids and to predict the region of interest. If we don't batch fold the box centroids, a given image could pass a unique identifier signal into bcentroids, allowing the network to then retrieve the region of interest on that image trivially. The remaining code computes proximity heatmaps based on the calculated (folded) box centroids, and then applies those heatmaps as masks in order to separate the predicted foreground and background regions. Those separated regions can then be used to make class predictions. Hope that helps!

**Comment by kehuantiantang** [comment]
I got it, thank you so much !
